e2e: node: re-enable the device plugin tests

ffromani opened the pull request:
#### What type of PR is this?

/kind cleanup
/kind failing-test

#### What this PR does / why we need it:
(fix if needed and) re-enable the device plugin tests

#### Which issue(s) this PR fixes:
Fixes #106635

#### Special notes for your reviewer:
Possibly already solved by other changes already committed, but we don't know yet because https://github.com/kubernetes/kubernetes/issues/106635#issuecomment-1105627265

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

ffromani committed:
e2e: node: re-enable the device plugin tests

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/sig node

fromanirh commented:
/test

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-swap-ubuntu-serial
/test pull-kubernetes-node-swap-fedora-serial

fromanirh commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

fromanirh commented:
nice, so https://github.com/kubernetes/kubernetes/issues/106635 reproduces on CI but not locally. I'll be investigating.

fromanirh commented:
/test pull-kubernetes-node-swap-fedora-serial
retrying job failed for apparently unrelated reasons (or just too early :) )

ffromani committed:
e2e: node: deviceplug: deepcopy the pod dev template

Let's avoid unexpected side effects

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd
using this lane as pilot. Failures in the other lanes all seem to share the same root cause at glance

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
initial findings about "Keeps device plugin assignments after the device plugin has been re-registered"
- the tests want to check the correct behaviour persisting the kubelet restart
- everything works well until the kubelet restarts
- during the kubelet restart, the device plugin stays running, but when kubelet is restarting, the kubelet endpoint becomes unavailable (obviously) and the device plugin panics, and stays in a wrong state
- after the restart, the test want to allocate another device, and this fails because the device plugin is in corrupted state

fromanirh commented:
the likely culprit is we still use the sample-device-plugin 1.3 image.

fromanirh commented:
> the likely culprit is we still use the sample-device-plugin 1.3 image.

Turns out it is not (just) this. After some more staring at the logs, this is the main issue:
kubelet:
E0512 13:23:14.363928   16940 client.go:64] "Unable to connect to device plugin client with socket path" err="failed to dial device plugin: context deadline exceeded" path="/var/lib/kubelet/device-plugins/dp.1652376184"
device plugin:
I0512 17:23:39.933131       1 stub.go:142] "Starting to serve on socket" socket="/var/lib/kubelet/plugins_registry//dp.1652376219"

```
@@ -64,7 +63,7 @@ var (
 var _ = SIGDescribe("Device Plugin [Feature:DevicePluginProbe][NodeFeature:DevicePluginProbe][Serial]", func() {
        f := framework.NewDefaultFramework("device-plugin-errors")
        f.NamespacePodSecurityEnforceLevel = admissionapi.LevelPrivileged
-       testDevicePlugin(f, "/var/lib/kubelet/plugins_registry")
+       testDevicePlugin(f, "/var/lib/kubelet/device-plugins")
 })
```

ffromani committed:
e2e: node: fix plugins directory

Previously, the e2e test was overriding the plugins socket directory to
"/var/lib/kubelet/plugins_registry". This seems wrong, and with that
setting the e2e test was already failing, because the registration
process was timing out, in turn because the kubelet was trying to call
back the device plugin in the wrong place (see below for details).

I can't explain why it worked before - or it if worked at all - but
it really seems that `pluginapi.DevicePluginPath` is the right
setting here.

+++

In a nutshell, the device plugin registration process works like this:

1. The kubelet runs and creates the device plugin socket registration
   endpoint:
	KubeletSocket = DevicePluginPath + "kubelet.sock"
	DevicePluginPath = "/var/lib/kubelet/device-plugins/"
2. Each device plugin will listen to an ENDPOINT the kubelet will connect
   backk to.  IOW the kubelet will act like a client to each device plugin,
   to perform allocation requests (and more)
   Each device plugin will serve from a endpoint.
   The endpoint name is plugin-specific, but they all must be inside a
   well-known directory: pluginapi.DevicePluginPath
3. The kubelet creates the device plugin pod, like any other pod
4. During the startup, each device plugin wants to register itself in the
   kubelet. So it sends a request through
   the registration endpoint. Key details:
	grpc.Dial(kubelet registration socket)
	registration request
	reqt := &pluginapi.RegisterRequest{
		Version:      pluginapi.Version,
		Endpoint:     endpointSocket,	<- socket relative to pluginapi.DevicePluginPath
		ResourceName: resourceName, 	<- resource name to be exposed
}
5. While handling the registration request, kubelet dial back the
   device plugin on socketDir + req.Endpoint.
   But socketDir is hardcoded in the device manager code to
   pluginapi.KubeletSocket

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-swap-ubuntu-serial
/test pull-kubernetes-node-swap-fedora-serial

fromanirh commented:
/retest pull-kubernetes-e2e-gce-ubuntu-containerd

fromanirh commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

fromanirh commented:
this is improving! now we are at
```
W0513 07:17:14.022] Latency metrics for node n1-standard-2-cos-89-16108-659-8-388be6cc
W0513 07:17:14.023] May 13 07:11:12.872: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
W0513 07:17:14.023] May 13 07:11:12.875: INFO: Condition Ready of node n1-standard-2-cos-89-16108-659-8-388be6cc is false instead of true. Reason: KubeletNotReady, message: PLEG is not healthy: pleg has yet to be successful
W0513 07:17:14.023] May 13 07:11:14.880: INFO: Condition Ready of node n1-standard-2-cos-89-16108-659-8-388be6cc is false instead of true. Reason: KubeletNotReady, message: PLEG is not healthy: pleg has yet to be successful
W0513 07:17:14.023] May 13 07:11:16.879: INFO: Condition Ready of node n1-standard-2-cos-89-16108-659-8-388be6cc is false instead of true. Reason: KubeletNotReady, message: PLEG is not healthy: pleg has yet to be successful
W0513 07:17:14.023] May 13 07:11:18.879: INFO: Condition Ready of node n1-standard-2-cos-89-16108-659-8-388be6cc is false instead of true. Reason: KubeletNotReady, message: PLEG is not healthy: pleg has yet to be successful
W0513 07:17:14.024] May 13 07:11:20.880: INFO: Condition Ready of node n1-standard-2-cos-89-16108-659-8-388be6cc is false instead of true. Reason: KubeletNotReady, message: PLEG is not healthy: pleg has yet to be successful
W0513 07:17:14.024] [1mSTEP[0m: Destroying namespace "device-plugin-errors-2182" for this suite.
W0513 07:17:14.024] 
W0513 07:17:14.024] [91m[1mâ€¢ Failure in Spec Teardown (AfterEach) [148.578 seconds][0m
W0513 07:17:14.024] [sig-node] Device Plugin [Feature:DevicePluginProbe][NodeFeature:DevicePluginProbe][Serial]
W0513 07:17:14.024] [90mtest/e2e_node/framework.go:23[0m
W0513 07:17:14.024]   DevicePlugin [Serial] [Disruptive]
W0513 07:17:14.024]   [90mtest/e2e_node/device_plugin_test.go:92[0m
W0513 07:17:14.025]     [91m[1mKeeps device plugin assignments after the device plugin has been re-registered [AfterEach][0m
W0513 07:17:14.025]     [90mtest/e2e_node/device_plugin_test.go:257[0m
W0513 07:17:14.025] 
W0513 07:17:14.025]     [91mUnexpected error:
W0513 07:17:14.025]         <*errors.errorString | 0xc00169e850>: {
W0513 07:17:14.025]             s: "there are currently no ready, schedulable nodes in the cluster",
W0513 07:17:14.025]         }
W0513 07:17:14.025]         there are currently no ready, schedulable nodes in the cluster
W0513 07:17:14.025]     occurred[0m
```
let's try the usual fix (longer timeout)

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
/hold
just in case

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
/test pull-kubernetes-unit

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
the last (?) problem is on `AfterEach` we do a last kubelet restart to clean up the node object from the sample resource which are no longer backed by the device plugin. It seems that randomly (but often) the node is slow to go back ready (PLEG not running), which makes a check fail deep down `getLocalNode` which bubbles up up to the top and makes the `AfterEach` fail at the very last stage.

I'm not sure why this started to fail lately, and I'm considering reimplementing this check tolerating this behaviour. Keep digging' because I don't want to hide bugs.

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-swap-ubuntu-serial
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-conformance-kind-ipv6-parallel

fromanirh commented:
> the last (?) problem is on `AfterEach` we do a last kubelet restart to clean up the node object from the sample resource which are no longer backed by the device plugin. It seems that randomly (but often) the node is slow to go back ready (PLEG not running), which makes a check fail deep down `getLocalNode` which bubbles up up to the top and makes the `AfterEach` fail at the very last stage.
> 
> I'm not sure why this started to fail lately, and I'm considering reimplementing this check tolerating this behaviour. Keep digging' because I don't want to hide bugs.

The latest change made the test pass without shortcut, but by tolerating a flipping in the kubelet ready state.
The gist of the issue is that `getLocalNode` doesn't tolerate flipping in the node ready condition. If it is called when the node is not ready, it make the test abort (`framework.ExpectNoError`) but this is too strict in this flow.

Now, for full disclosure I can't fully explain why the ready state flip/flops, but for the purpose of this specific test we can tolerate this condition without handwaving this behaviour and while keep being correct.

```
I0513 20:23:46.088] [1mSTEP[0m: Waiting for devices to become unavailable on the local node
I0513 20:23:46.088] May 13 20:15:39.582: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" unschedulable=false ready=true
I0513 20:23:46.089] May 13 20:15:39.582: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" conditions=[]v1.NodeCondition{v1.NodeCondition{Type:"MemoryPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 14, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientMemory", Message:"kubelet has sufficient memory available"}, v1.NodeCondition{Type:"DiskPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 14, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 52, 12, 0, time.Local), Reason:"KubeletHasNoDiskPressure", Message:"kubelet has no disk pressure"}, v1.NodeCondition{Type:"PIDPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 14, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientPID", Message:"kubelet has sufficient PID available"}, v1.NodeCondition{Type:"Ready", Status:"True", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 14, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 20, 13, 49, 0, time.Local), Reason:"KubeletReady", Message:"kubelet is posting ready status. AppArmor enabled"}}
I0513 20:23:46.090] May 13 20:15:39.582: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" numOfSampleResources=2
I0513 20:23:46.090] May 13 20:15:41.585: INFO: Condition Ready of node n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e is false instead of true. Reason: KubeletNotReady, message: container runtime status check may not have completed yet
I0513 20:23:46.090] May 13 20:15:41.585: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" unschedulable=false ready=false
I0513 20:23:46.091] May 13 20:15:41.585: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" conditions=[]v1.NodeCondition{v1.NodeCondition{Type:"MemoryPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientMemory", Message:"kubelet has sufficient memory available"}, v1.NodeCondition{Type:"DiskPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 52, 12, 0, time.Local), Reason:"KubeletHasNoDiskPressure", Message:"kubelet has no disk pressure"}, v1.NodeCondition{Type:"PIDPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientPID", Message:"kubelet has sufficient PID available"}, v1.NodeCondition{Type:"Ready", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), Reason:"KubeletNotReady", Message:"container runtime status check may not have completed yet"}}
I0513 20:23:46.091] May 13 20:15:41.585: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" numOfSampleResources=2
I0513 20:23:46.092] May 13 20:15:43.591: INFO: Condition Ready of node n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e is false instead of true. Reason: KubeletNotReady, message: container runtime status check may not have completed yet
I0513 20:23:46.092] May 13 20:15:43.591: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" unschedulable=false ready=false
I0513 20:23:46.093] May 13 20:15:43.591: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" conditions=[]v1.NodeCondition{v1.NodeCondition{Type:"MemoryPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientMemory", Message:"kubelet has sufficient memory available"}, v1.NodeCondition{Type:"DiskPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 52, 12, 0, time.Local), Reason:"KubeletHasNoDiskPressure", Message:"kubelet has no disk pressure"}, v1.NodeCondition{Type:"PIDPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientPID", Message:"kubelet has sufficient PID available"}, v1.NodeCondition{Type:"Ready", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), Reason:"KubeletNotReady", Message:"container runtime status check may not have completed yet"}}
I0513 20:23:46.093] May 13 20:15:43.591: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" numOfSampleResources=2
I0513 20:23:46.093] May 13 20:15:45.596: INFO: Condition Ready of node n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e is false instead of true. Reason: KubeletNotReady, message: container runtime status check may not have completed yet
I0513 20:23:46.093] May 13 20:15:45.596: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" unschedulable=false ready=false
I0513 20:23:46.094] May 13 20:15:45.596: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" conditions=[]v1.NodeCondition{v1.NodeCondition{Type:"MemoryPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientMemory", Message:"kubelet has sufficient memory available"}, v1.NodeCondition{Type:"DiskPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 52, 12, 0, time.Local), Reason:"KubeletHasNoDiskPressure", Message:"kubelet has no disk pressure"}, v1.NodeCondition{Type:"PIDPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientPID", Message:"kubelet has sufficient PID available"}, v1.NodeCondition{Type:"Ready", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), Reason:"KubeletNotReady", Message:"container runtime status check may not have completed yet"}}
I0513 20:23:46.095] May 13 20:15:45.596: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" numOfSampleResources=2
I0513 20:23:46.095] May 13 20:15:47.602: INFO: Condition Ready of node n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e is false instead of true. Reason: KubeletNotReady, message: container runtime status check may not have completed yet
I0513 20:23:46.095] May 13 20:15:47.602: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" unschedulable=false ready=false
I0513 20:23:46.096] May 13 20:15:47.602: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" conditions=[]v1.NodeCondition{v1.NodeCondition{Type:"MemoryPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientMemory", Message:"kubelet has sufficient memory available"}, v1.NodeCondition{Type:"DiskPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 52, 12, 0, time.Local), Reason:"KubeletHasNoDiskPressure", Message:"kubelet has no disk pressure"}, v1.NodeCondition{Type:"PIDPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientPID", Message:"kubelet has sufficient PID available"}, v1.NodeCondition{Type:"Ready", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), Reason:"KubeletNotReady", Message:"container runtime status check may not have completed yet"}}
I0513 20:23:46.096] May 13 20:15:47.602: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" numOfSampleResources=2
I0513 20:23:46.097] May 13 20:15:49.605: INFO: Condition Ready of node n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e is false instead of true. Reason: KubeletNotReady, message: container runtime status check may not have completed yet
I0513 20:23:46.097] May 13 20:15:49.606: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" unschedulable=false ready=false
I0513 20:23:46.098] May 13 20:15:49.606: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" conditions=[]v1.NodeCondition{v1.NodeCondition{Type:"MemoryPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientMemory", Message:"kubelet has sufficient memory available"}, v1.NodeCondition{Type:"DiskPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 52, 12, 0, time.Local), Reason:"KubeletHasNoDiskPressure", Message:"kubelet has no disk pressure"}, v1.NodeCondition{Type:"PIDPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientPID", Message:"kubelet has sufficient PID available"}, v1.NodeCondition{Type:"Ready", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), Reason:"KubeletNotReady", Message:"container runtime status check may not have completed yet"}}
I0513 20:23:46.098] May 13 20:15:49.606: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" numOfSampleResources=2
I0513 20:23:46.098] May 13 20:15:51.610: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" unschedulable=false ready=true
I0513 20:23:46.099] May 13 20:15:51.610: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" conditions=[]v1.NodeCondition{v1.NodeCondition{Type:"MemoryPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientMemory", Message:"kubelet has sufficient memory available"}, v1.NodeCondition{Type:"DiskPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 52, 12, 0, time.Local), Reason:"KubeletHasNoDiskPressure", Message:"kubelet has no disk pressure"}, v1.NodeCondition{Type:"PIDPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientPID", Message:"kubelet has sufficient PID available"}, v1.NodeCondition{Type:"Ready", Status:"True", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 20, 15, 50, 0, time.Local), Reason:"KubeletReady", Message:"kubelet is posting ready status. AppArmor enabled"}}
I0513 20:23:46.099] May 13 20:15:51.611: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" numOfSampleResources=0
I0513 20:23:46.100] [1mSTEP[0m: devices now unavailable on the local node
I0513 20:23:46.100] [AfterEach] [sig-node] Device Plugin [Feature:DevicePluginProbe][NodeFeature:DevicePluginProbe][Serial]
I0513 20:23:46.100]   test/e2e/framework/framework.go:187
```
I'll explain and elaborate in more detail in the final proposed patch

fromanirh commented:
/test pull-kubernetes-node-swap-ubuntu-serial
/test pull-kubernetes-node-swap-fedora-serial

ffromani committed:
e2e: node: devplugin: tolerate node readiness flip

In the AfterEach check of the e2e node device plugin tests,
the tests want really bad to clean up after themselves:
- delete the sample device plugin
- restart again the kubelet
- ensure that after the restart, no stale sample devices
  (provided  by the sample device plugin) are reported anymore.

We observed that in the AfterEach block of these e2e tests
we have quite reliably a flip/flop of the kubelet readiness
state, possibly related to a race with/ a slow runtime/PLEG check.

What happens is that the kubelet readiness state is true,
but goes false for a quick interval and then goes true again
and it's pretty stable after that (observed adding more logs
to the check loop).

The key factor here is the function `getLocalNode` aborts the
test (as in `framework.ExpectNoError`) if the node state is
not ready. So any occurrence of this scenario, even if it
is transient, will cause a test failure. I believe this will
make the e2e test unnecessarily fragile without making it more
correct.

For the purpose of the test we can tolerate this kind of glitches,
with kubelet flip/flopping the ready state, granted that we meet
eventually the final desired condition on which the node reports
ready AND reports no sample devices present - which was the condition
the code was trying to check.

So, we add a variant of `getLocalNode`, which just fetches the
node object the e2e_node framework created, alongside to a flag
reporting the node readiness. The new helper does not make
implicitly the test abort if the node is not ready, just bubbles
up this information.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
> > the last (?) problem is on `AfterEach` we do a last kubelet restart to clean up the node object from the sample resource which are no longer backed by the device plugin. It seems that randomly (but often) the node is slow to go back ready (PLEG not running), which makes a check fail deep down `getLocalNode` which bubbles up up to the top and makes the `AfterEach` fail at the very last stage.
> > I'm not sure why this started to fail lately, and I'm considering reimplementing this check tolerating this behaviour. Keep digging' because I don't want to hide bugs.
> 
> The latest change made the test pass without shortcut, but by tolerating a flipping in the kubelet ready state. The gist of the issue is that `getLocalNode` doesn't tolerate flipping in the node ready condition. If it is called when the node is not ready, it make the test abort (`framework.ExpectNoError`) but this is too strict in this flow.
> 
> Now, for full disclosure I can't fully explain why the ready state flip/flops, but for the purpose of this specific test we can tolerate this condition without handwaving this behaviour and while keep being correct.
> 
> ```
> I0513 20:23:46.088] �[1mSTEP�[0m: Waiting for devices to become unavailable on the local node
> I0513 20:23:46.088] May 13 20:15:39.582: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" unschedulable=false ready=true
> I0513 20:23:46.089] May 13 20:15:39.582: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" conditions=[]v1.NodeCondition{v1.NodeCondition{Type:"MemoryPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 14, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientMemory", Message:"kubelet has sufficient memory available"}, v1.NodeCondition{Type:"DiskPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 14, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 52, 12, 0, time.Local), Reason:"KubeletHasNoDiskPressure", Message:"kubelet has no disk pressure"}, v1.NodeCondition{Type:"PIDPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 14, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientPID", Message:"kubelet has sufficient PID available"}, v1.NodeCondition{Type:"Ready", Status:"True", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 14, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 20, 13, 49, 0, time.Local), Reason:"KubeletReady", Message:"kubelet is posting ready status. AppArmor enabled"}}
> I0513 20:23:46.090] May 13 20:15:39.582: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" numOfSampleResources=2
> I0513 20:23:46.090] May 13 20:15:41.585: INFO: Condition Ready of node n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e is false instead of true. Reason: KubeletNotReady, message: container runtime status check may not have completed yet
> I0513 20:23:46.090] May 13 20:15:41.585: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" unschedulable=false ready=false
> I0513 20:23:46.091] May 13 20:15:41.585: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" conditions=[]v1.NodeCondition{v1.NodeCondition{Type:"MemoryPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientMemory", Message:"kubelet has sufficient memory available"}, v1.NodeCondition{Type:"DiskPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 52, 12, 0, time.Local), Reason:"KubeletHasNoDiskPressure", Message:"kubelet has no disk pressure"}, v1.NodeCondition{Type:"PIDPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientPID", Message:"kubelet has sufficient PID available"}, v1.NodeCondition{Type:"Ready", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), Reason:"KubeletNotReady", Message:"container runtime status check may not have completed yet"}}
> I0513 20:23:46.091] May 13 20:15:41.585: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" numOfSampleResources=2
> I0513 20:23:46.092] May 13 20:15:43.591: INFO: Condition Ready of node n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e is false instead of true. Reason: KubeletNotReady, message: container runtime status check may not have completed yet
> I0513 20:23:46.092] May 13 20:15:43.591: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" unschedulable=false ready=false
> I0513 20:23:46.093] May 13 20:15:43.591: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" conditions=[]v1.NodeCondition{v1.NodeCondition{Type:"MemoryPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientMemory", Message:"kubelet has sufficient memory available"}, v1.NodeCondition{Type:"DiskPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 52, 12, 0, time.Local), Reason:"KubeletHasNoDiskPressure", Message:"kubelet has no disk pressure"}, v1.NodeCondition{Type:"PIDPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientPID", Message:"kubelet has sufficient PID available"}, v1.NodeCondition{Type:"Ready", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), Reason:"KubeletNotReady", Message:"container runtime status check may not have completed yet"}}
> I0513 20:23:46.093] May 13 20:15:43.591: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" numOfSampleResources=2
> I0513 20:23:46.093] May 13 20:15:45.596: INFO: Condition Ready of node n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e is false instead of true. Reason: KubeletNotReady, message: container runtime status check may not have completed yet
> I0513 20:23:46.093] May 13 20:15:45.596: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" unschedulable=false ready=false
> I0513 20:23:46.094] May 13 20:15:45.596: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" conditions=[]v1.NodeCondition{v1.NodeCondition{Type:"MemoryPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientMemory", Message:"kubelet has sufficient memory available"}, v1.NodeCondition{Type:"DiskPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 52, 12, 0, time.Local), Reason:"KubeletHasNoDiskPressure", Message:"kubelet has no disk pressure"}, v1.NodeCondition{Type:"PIDPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientPID", Message:"kubelet has sufficient PID available"}, v1.NodeCondition{Type:"Ready", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), Reason:"KubeletNotReady", Message:"container runtime status check may not have completed yet"}}
> I0513 20:23:46.095] May 13 20:15:45.596: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" numOfSampleResources=2
> I0513 20:23:46.095] May 13 20:15:47.602: INFO: Condition Ready of node n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e is false instead of true. Reason: KubeletNotReady, message: container runtime status check may not have completed yet
> I0513 20:23:46.095] May 13 20:15:47.602: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" unschedulable=false ready=false
> I0513 20:23:46.096] May 13 20:15:47.602: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" conditions=[]v1.NodeCondition{v1.NodeCondition{Type:"MemoryPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientMemory", Message:"kubelet has sufficient memory available"}, v1.NodeCondition{Type:"DiskPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 52, 12, 0, time.Local), Reason:"KubeletHasNoDiskPressure", Message:"kubelet has no disk pressure"}, v1.NodeCondition{Type:"PIDPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientPID", Message:"kubelet has sufficient PID available"}, v1.NodeCondition{Type:"Ready", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), Reason:"KubeletNotReady", Message:"container runtime status check may not have completed yet"}}
> I0513 20:23:46.096] May 13 20:15:47.602: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" numOfSampleResources=2
> I0513 20:23:46.097] May 13 20:15:49.605: INFO: Condition Ready of node n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e is false instead of true. Reason: KubeletNotReady, message: container runtime status check may not have completed yet
> I0513 20:23:46.097] May 13 20:15:49.606: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" unschedulable=false ready=false
> I0513 20:23:46.098] May 13 20:15:49.606: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" conditions=[]v1.NodeCondition{v1.NodeCondition{Type:"MemoryPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientMemory", Message:"kubelet has sufficient memory available"}, v1.NodeCondition{Type:"DiskPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 52, 12, 0, time.Local), Reason:"KubeletHasNoDiskPressure", Message:"kubelet has no disk pressure"}, v1.NodeCondition{Type:"PIDPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientPID", Message:"kubelet has sufficient PID available"}, v1.NodeCondition{Type:"Ready", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 20, 15, 39, 0, time.Local), Reason:"KubeletNotReady", Message:"container runtime status check may not have completed yet"}}
> I0513 20:23:46.098] May 13 20:15:49.606: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" numOfSampleResources=2
> I0513 20:23:46.098] May 13 20:15:51.610: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" unschedulable=false ready=true
> I0513 20:23:46.099] May 13 20:15:51.610: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" conditions=[]v1.NodeCondition{v1.NodeCondition{Type:"MemoryPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientMemory", Message:"kubelet has sufficient memory available"}, v1.NodeCondition{Type:"DiskPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 52, 12, 0, time.Local), Reason:"KubeletHasNoDiskPressure", Message:"kubelet has no disk pressure"}, v1.NodeCondition{Type:"PIDPressure", Status:"False", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 19, 15, 8, 0, time.Local), Reason:"KubeletHasSufficientPID", Message:"kubelet has sufficient PID available"}, v1.NodeCondition{Type:"Ready", Status:"True", LastHeartbeatTime:time.Date(2022, time.May, 13, 20, 15, 50, 0, time.Local), LastTransitionTime:time.Date(2022, time.May, 13, 20, 15, 50, 0, time.Local), Reason:"KubeletReady", Message:"kubelet is posting ready status. AppArmor enabled"}}
> I0513 20:23:46.099] May 13 20:15:51.611: INFO: node "n1-standard-2-ubuntu-gke-2004-1-20-v20220511-b6ddea3e" numOfSampleResources=0
> I0513 20:23:46.100] �[1mSTEP�[0m: devices now unavailable on the local node
> I0513 20:23:46.100] [AfterEach] [sig-node] Device Plugin [Feature:DevicePluginProbe][NodeFeature:DevicePluginProbe][Serial]
> I0513 20:23:46.100]   test/e2e/framework/framework.go:187
> ```
> 
> I'll explain and elaborate in more detail in the final proposed patch

I'm happy enough with the fix added in the last upload. Now I need to figure out the root cause of https://github.com/kubernetes/kubernetes/pull/109820#issuecomment-1125263178, then I'll lift the WIPs and the PR will be ready at last.

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
```
invalid_grant: Invalid JWT Signature.
```
:\

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2

fromanirh commented:
/test pull-kubernetes-node-swap-ubuntu-serial

fromanirh commented:
/test pull-kubernetes-node-swap-fedora-serial

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
/test pull-kubernetes-node-swap-fedora-serial

ffromani committed:
e2e: node: reorder imports

trivial cleanup

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/hold cancel
/remove-sig testing
/remove-label area/e2e-test-framework

fromanirh commented:
/remove-area e2e-test-framework

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
pr ready for review!

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-containerd

fromanirh commented:
/test pull-kubernetes-node-swap-fedora-serial

fromanirh commented:
```
I0518 08:38:22.226] unable to create gce instance with running docker daemon for image fedora-coreos-35-20220424-3-0-gcp-x86-64.  instance n1-standard-2-fedora-coreos-35-20220424-3-0-gcp-x86-64-b9a9663e not running containerd/crio daemon - Command failed: ssh: connect to host 34.145.36.136 port 22: Connection refused
```

endocrimes commented:
/test pull-kubernetes-node-swap-fedora-serial

fromanirh commented:
something seems wrong with the lane:
```
 W0522 13:33:51.325] ERROR: (gcloud.compute.addresses.describe) Could not fetch resource:
W0522 13:33:51.325]  - The resource 'projects/k8s-infra-e2e-boskos-110/regions/us-west1/addresses/bootstrap-e2e-master-ip' was not found 
```

SergeyKanzhelev commented:
/triage accepted
/priority important-soon
/assign
